packages/flutter/lib/src/painting/_network_image_io.dart

```diff
@@ -81,39 +73,115 @@ class NetworkImage extends image_provider.ImageProvider<image_provider.NetworkIm
       assert(key == this);
 
       final Uri resolved = Uri.base.resolve(key.url);
-      final HttpClientRequest request = await _httpClient.getUrl(resolved);
-      headers?.forEach((String name, String value) {
-        request.headers.add(name, value);
+
+      final Completer<TransferableTypedData> bytesCompleter = Completer<TransferableTypedData>();
+      final RawReceivePort downloadResponseHandler = RawReceivePort((_DownloadResponse response) {
+        if (response.bytes != null) {
+          if (bytesCompleter.isCompleted) {
+            // If an uncaught error occurred in the worker isolate, we'll have
+            // already completed our bytes completer.
+            return;
+          }
+          bytesCompleter.complete(response.bytes);
+        } else if (response.chunkEvent != null) {
+          chunkEvents.add(response.chunkEvent);
+        } else if (response.error != null) {
+          bytesCompleter.completeError(response.error);
+        } else {
+          assert(false);
+        }
       });
-      final HttpClientResponse response = await request.close();
-      if (response.statusCode != HttpStatus.ok)
-        throw Exception('HTTP request failed, statusCode: ${response?.statusCode}, $resolved');
-
-      final Uint8List bytes = await consolidateHttpClientResponseBytes(
-        response,
-        onBytesReceived: (int cumulative, int total) {
-          chunkEvents.add(ImageChunkEvent(
-            cumulativeBytesLoaded: cumulative,
-            expectedTotalBytes: total,
-          ));
-        },
-      );
-      if (bytes.lengthInBytes == 0)
+
+      // This will keep reference to [debugNetworkImageHttpClientProvider] tree-shaken
+      // out of release builds.
+      HttpClientProvider httpClientProvider;
+      assert(() { httpClientProvider = debugNetworkImageHttpClientProvider; return true; }());
+
+      final _DownloadRequest downloadRequest = _DownloadRequest(downloadResponseHandler.sendPort,
+          resolved, headers, httpClientProvider);
+      if (_requestPort != null) {
+        // If worker isolate is properly set up ([_requestPort] is holding
+        // initialized [SendPort]), then just send download request down to it.
+        _requestPort.send(downloadRequest);
+      } else {
+        if (_pendingLoader == null) {
+          // If worker isolate creation was not started, start creation now.
+          assert(_pendingLoadRequests == null);
+          _pendingLoadRequests = <_PendingLoadRequest>[];
+          _pendingLoader = _setupIsolate()..then((Isolate isolate) {
+              final RawReceivePort handleError = RawReceivePort((List<dynamic> errorAndStackTrace) {
+                _cleanupDueToError(errorAndStackTrace[0]);
```

Comment:
It feels weird that upon any uncaught exception in the isolate, we'll set `_pendingLoader` to null and thus lose our reference to the worker isolate, spawning a new isolate if another request comes in.

Further, it currently only completes the _pending_ requests with an error, but those pending requests get cleared as soon as the worker isolate starts up, so generally (always?), it'll be empty.

What if we wired up uncaught exceptions in the worker isolate to `FlutterError.reportError()`?

Reply:
Not sure I see the problem with current setup.
_pendingLoader/_pendingLoadRequests are there to handle initial load during isolate creation. They won't be used once isolate is created and provided it's send port.